Docker stack

Pourbaix opened the pull request:
## Pull requests for docker-stack script

Adding a shell script that allows devs to initialize a version of Dspace with some users, collections and groups already created.

renaudmichotte committed:
[ADD] feat (scripts): creating first setup script

Creating the bash setup script use to import sample data into fresh and clean DSpace install.

Co-authored-by: Renaud Michotte <renaud.michotte@uclouvain.be>

Commit:
[ADD] feat(init): Adding backend stack + init script

zannkukai commented on .gitignore:
```diff
@@ -49,3 +49,10 @@ rebel.xml
+__pychache__
+
+## Ignore logs files
+.log
```
Change to "log/*" here. A log file is present into your PR

zannkukai commented on dspace/config/init-sample.xml:
```diff
@@ -0,0 +1,21 @@
+		<description />
+		<intro />
+		<copyright />
+		<sidebar />
```
Is these tags are really required (even if empty). If not, remote them.

Pourbaix replied:
In fact, it seems not to be required: https://wiki.lyrasis.org/display/DSDOC7x/Exporting+and+Importing+Community+and+Collection+Hierarchy

zannkukai commented on dspace/config/init-sample.xml:
```diff
@@ -0,0 +1,21 @@
+			<name>Master thesis</name>
+			<entity-type>Thesis</entity-type>
+			<submission-type>master-thesis</submission-type>
+			<description />
+			<intro />
+			<copyright />
+			<sidebar />
+			<license />
+			<provenance />
```
same here. Try to remove unnecessary tags

Pourbaix replied:
See first one

zannkukai commented on scripts/setup/__init__.py:
Setup isn't a python package file, just a simple directory containing python scripts. Remove this `__init__.py` file

zannkukai commented on scripts/setup/__pycache__/dspaceAPI.cpython-311.pyc:
remove all "*.pyc" from repository.
You could include this default lines into `.gitignore` file

```
# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]

# C extensions
*.so

# Distribution / packaging
bin/
build/
develop-eggs/
dist/
eggs/
lib/
lib64/
parts/
sdist/
var/
*.egg-info/
.installed.cfg
*.egg

# Installer logs
pip-log.txt
pip-delete-this-directory.txt

# Unit test / coverage reports
.tox/
.coverage
.cache
nosetests.xml
coverage.xml

# Translations
*.mo
```

zannkukai commented on scripts/setup/log/2023-10-30_16:19:33.log:
remove this file from your PR/repository

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+    sys.exit(1)
+
+## Recover the name of the module for the main logger
+logger = logging.getLogger(__name__)
```
place this at the top of the file (near constant declaration)

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+# implements all the HTTP calls we need.
+#########################################
+
+def wrong(message):
```
Should be more useful and generic
```python
def wrong(message=None, exception=None, exit_code=1):
    """Log an error message and/or the corresponding exception
    :param message: the error message to log.
    :param exception: the ``Exception`` to capture into log.
    :param exit_code: the exit code to return
    """
    assert message or exception, "at least 'message' or 'exception' parameter is required"
    logger.error(message or str(exception))
    if exception:
        logger.exception(exception)
    sys.exit(exit_code)

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+except Exception as e:
+    logger.error("❌ An error occured while loading the configuration file ❌")
+    logger.error(e)
+    sys.exit(1)
```
use `wrong()` method now

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+res = api.acceptUserAgreement(auth_token)
+logger.info("✅ Accepted admin's users agreement")
+
+
```
Too many blank lines for PEP-8. We could not become PEP-9 guru, but keep best practices in our minds.

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+    logger.error(e)
+    sys.exit(1)
+
+# LOGIN AS ADMIN AND RETREIVE TOKEN
```
Maybe add more docstring to explain how and why you need to do these operations, in this order.
Something like
```
# LOGIN AS ADMINISTATOR
#    We will use the Backend API to login as an administrator to next perform multiple items creations.
#    To have full access, we also need mark this account with "user agreement" check mark ; after this operation
#    we need to refresh to first receive token to have a new one for full access management.
```
My english is far as perfect... You can check with @manadeweerdt for correction

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+#  Accept user agreement for other users
+for user in users:
+    res = api.acceptUserAgreement(auth_token, user["email"])
+logger.info("✅ Accepted alt users agreement")
```
`alt` or `all` ?

Pourbaix replied:
Alt like in "alternative", so basically users, librarian and manager in this case.

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+        collections = config["collections"]
+        logger.debug("✅ Config loaded")
+except Exception as e:
+    logger.error("❌ An error occured while loading the configuration file ❌")
```
only one "❌" at the beginning of the line

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+    auth_token = auth_res.headers["Authorization"]
+    logger.info("✅ Connected to Admin account")
+else:
+    wrong("❌ Could not connect with the given admin email and password ❌")
```
one "❌"

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+    # Adding metadata schemas 
+    res = api.addMetadataSchema(schema["prefix"], schema["namespace"], auth_token)
+    if not res.ok:
+        wrong("❌ An error occured while adding metadata schema: %s" % schema["prefix"])
```
We will deal with python 3.X, we will use the "f''" syntax
```python
wrong(f"❌ An error occurred while adding metadata schema: {schema['prefix']}")
```

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+        # Adding metadata fields for the current schema
+        res = api.addMetadataField(str(schema_id), auth_token, field["element"], field["qualifier"], field["scope_note"])
+        if not res.ok:
+            wrong("❌ An error occured while adding metadata field to %s" % str(schema_id))
```
use `f()` string
Could be great to have a debug message if response is ok to know which schemas is successfully loaded

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+for group in eperson_groups:
+    res = api.createEpersonGroup(auth_token, group["name"], group["description"])
+    if not res.ok:
+        wrong("❌ An error occured while creating group %s" % group["name"])
```
same again

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+
+    group_info = res.json()
+    group_id = group_info["id"]
+
```
remove blank line when it isn't necessary

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+        user_detail = api.getEpersonFromEmail(auth_token, user["email"]).json()
+        user_id_path = user_detail["_links"]["self"]["href"]
+        res = api.addEpersonToGroup(user_id_path, group_id, auth_token)
+        if not res.ok:
```
debug message if "ok"

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+                wrong("Error while linking child group %s to parent %s" % (group, workflow_group["id"]))
+    logger.info("✅ Added %s collection rights" % collection["name"])
+
+sys.exit(0)
```
is this relevant ? Do you check this return statement or "0" isn't the default code ?

Pourbaix replied:
In fact, it seems like python is executing with code 0 by default when it has finished executing a script.

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+        user_id_path = user_detail["_links"]["self"]["href"]
+        res = api.addEpersonToGroup(user_id_path, group_id, auth_token)
+        if not res.ok:
+            wrong("❌ An error occured while adding the user %s to the %s group" % (user["email"], group["name"]))
```
`when` instead of `while` (check with @manadeweerdt this is our english guru)

zannkukai commented on scripts/setup/setup.py:
```diff
@@ -0,0 +1,137 @@
+            
+            res = api.addSubGroupToGroup(workflow_group["id"], sub_group_uri, auth_token)
+            if not res.ok:
+                wrong("Error while linking child group %s to parent %s" % (group, workflow_group["id"]))
```
missing "❌" and use the "f''" synhtax

zannkukai commented on scripts/setup/setup.py:
As general comment : 
 * use `f"string {variable}"` to format string since this is new and preferred way in Python 3.X
 * use single quote `'` instead of double quote `"` around strings when it's possible (recommendation of PEP-8)
 * To to add "Sourcery" plugin on VSCode to help code refactoring

zannkukai commented on scripts/setup/setup.sh:
```diff
@@ -0,0 +1,128 @@
+    echo "✅ Found the container, performing actions ..."
+
+    # STEP ONE: CLEAN DATABASE
+
```
COSMETIC : to to limit blank lines to have better visual impact and have better block-code delimitation
Use number (better visual impact) than string representation for number. Maybe "STEP #1 :: Clean database"

zannkukai commented on scripts/setup/setup.sh:
```diff
@@ -0,0 +1,128 @@
+    
+    echo "👤 Creating user groups, metadata registers and managing collection rights..."
+
+    python3 ${WORKING_PATH}/setup.py >> ${LOG_PATH}
```
simple "python" instead of "python3".
include a "requirement.txt" file and specify than python at least 3.X (3.10) is required

Pourbaix replied:
'python' links to python2, normally to use python 3.X version you need to type 'python3' as described in this source:
https://stackoverflow.com/questions/67550189/i-dont-have-python-but-i-have-python3 .
Apparently, you could also install 'python-is-python3' with apt (or any other package manager) to link the 'python' command to the actual python 3.X interpreter.

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+    uuid: str 
+        The uuid of the current logged in user
+    """
+    def __init__(self, url):
```
docstring for this function ; specially for "params"

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+            "editor": "/workflowGroups/editor", 
+            "finaleditor": "/workflowGroups/finaleditor", 
+            "bitstreamread": "/bitstreamReadGroup",
+            "itemread": "/itemReadGroup"
```
use single quote `'` instead of `"` when it's possible

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+        -------------------------------------------------
+        Generate a header dictionnary that can be used to make request while beeing authenticated.
+        
+        :return:(Response Object) The response of the request made on the base API url. 
```
With python3, we could use "typing" to specify params type and return statement.
Always add `:param` docstring for all param of the function.

According PEP-8, the docstring must declare:
  * first line : purpose of the function in MAX one line. Finishing by a point "."
  * second line : blank line
  * one line (or more) by param (exception for "self" of course)
  * use ":param", ":return", ":raise"

```python
def __generateAuthHeaders(self, auth_token: str) -> dict:
    """Generate authorization headers required for DSpace API call.

     :param auth_token: the authorization token to include into header.
     :return: headers to use into a `request.Request` as a dictionary.
     :raise SampleAndCustomError: When_my_custom_exception occured.
     """
```

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+        """
+        res = self.httpSession.get(self.url)
+        if res.status_code != 200:
+            raise Exception("Error while calling the api url, got status code " + res.status_code + " !")
```
Avoid to raise generic "Exception". Here you could raise an "HTTPError" exception (or custom exception inherits from `HTTPError` to be hardcore_and_overkill)

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+        
+        :return:(str) The XSRF Token
+        """
+        self.getRoot()
```
is it required ? Seems not used

Pourbaix replied:
Required to refresh the cookies

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+        }
+        route = f"/authn/login?user={data['user']}&password={data['password']}"
+        res = self.httpSession.post(self.url + route, data=data, headers=headers, cookies=self.cookies)
+        if res.status_code == 200:
```
COSMETIC : use "request.codes.ok" instead of "200"... it's more generic

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+    
+    def refreshConnectionWithToken(self, auth_token):
+        headers = self.__generateAuthHeaders(auth_token)
+
```
remove unnecessary blank lines

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+        if email is None:
+            if not self.user_email:
+                raise Exception("Error: Authenticate a user before retreiving current user informations")
+            target_email = self.user_email
```
```python
target_email = email or self.user_email
if not target_email:
    raise CustomException("Authenticate a user before retr**ie**ving current user informations")
```

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+    #     }
+    #     res_pwd = self.patchEpersonInfo(res.json()["id"], [operation], auth_token)
+    #     print(res_pwd.text)
+    #     return res
```
Remove if not necessary now or in future.

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+                    raise Exception("Error while patching user informations")
+        except Exception as e:
+            raise Exception("Error while extracting the user metadata:\n" + str(e))
+        return 0
```
Wrong according docstring

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+                    "op": "add",
+                    "path": "/metadata/dspace.agreements.end-user",
+                    "value": [
+                        {"value": "true"},
```
`{"value": True},`

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+
+        groups_list = self.getAllGroups(auth_token).json()["_embedded"]["groups"]
+
+        target_group = next(filter(lambda x: x["name"] == group_name, groups_list))
+        return target_group
```
single line : `return next(filter(lambda x: x["name"] == group_name, groups_list))`

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+            }
+        }
+
+        res = self.httpSession.post(self.url + "/eperson/groups", json=obj, headers=headers)
+        return res
+
```
single line

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+        headers = self.__generateAuthHeaders(auth_token)        
+        headers["Content-Type"] = "text/uri-list"
+
+        res = self.httpSession.post(self.url + "/eperson/groups/" + group_id + "/epersons", data=erperson_uri, headers=headers)
+        return res
```
same

zannkukai commented on scripts/setup/dspaceAPI.py:
```diff
@@ -0,0 +1,499 @@
+        """
+        Create a resource policy in the Dspace App and grant it to a group or a user.
+        """
+        if (epersonUUID and groupUUID):
```
remove unnecessary parentheses (Sourcery plugin is a great help in this case)

zannkukai commented on scripts/setup/config/setup.json:
maybe it could be great to split this single file into multiple smaller file : "users.json", "groups.json", ...

Commit:
[UPD] feat(setup script): Refactoring a lot of code for PR
-> Better docstring
-> Added typing
-> Added custom Exceptions
-> Some logic refactoring
-> Divided config files in 4 (one per type)

Commit:
[DEL] feat(logs): Removing tracking for logs files

Commit:
[UPD] feat(metadata-schemas): Adding new metadata schema: tags, authors and advisors
